Auto commit disabled by default

sushantdhiman opened the pull request:
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?
- [x] Have you added new tests to prevent regression ?

### Description of change

`AUTO COMMIT` query is not sent by default except for `MySQL`. Closes #5094

mickhansen commented:
I'd love a unit test or two for this. Simply starting a transaction and verifying it runs the expected SQL queries.

It could be the same unit tests covering transaction preperation in general, assuming we make other changes.

sushantdhiman commented:
Added the unit tests. From unit test I found out `sqlite` was wasting one query by sending an effective comment in place of auto commit query. Fixed it as well.

mickhansen commented on test/unit/transaction.test.js:
```diff
@@ -0,0 +1,57 @@
+        expect(sqlIndex.join(' ').indexOf('SET autocommit = 1;')).to.be.eql(-1);
+        return Promise.resolve({});
+      });
+    });
```
Could we test on the sequence instead, test again an array of SQL calls.

sushantdhiman replied:
I can use the latest `sql` statement from `query` stub. That will have the `auto commit`

mickhansen commented on test/unit/transaction.test.js:
```diff
@@ -0,0 +1,57 @@
+
+  before(function () {
+    this.stub = sinon.stub(current, 'query', function (sql) {
+      sqlIndex.push(sql);
```
Easier to just look at the args for the stub, no need to maintain this.

sushantdhiman replied:
Right

mickhansen commented on test/unit/transaction.test.js:
```diff
@@ -0,0 +1,57 @@
+  var sqlIndex = [];
+
+  before(function () {
+    this.stub = sinon.stub(current, 'query', function (sql) {
```
sinon.stub(current, 'query').returns(Promise.resolve({}));

sushantdhiman replied:
Was experimenting a bit earlier, forgot to change it.

sushantdhiman commented:
Refactored tests as per your comments @mickhansen

mickhansen commented on test/unit/transaction.test.js:
```diff
@@ -0,0 +1,53 @@
+        return Promise.resolve({});
+      });
+    });
+  }
```
Sorry for bikeshedding, but perhaps something like:

```
var expectations = {
  all: [
    'START TRANSACTION',
    'COMMIT'
  ],
  mysql: [

   ]
};
return current.transaction(function(t) {
  return Promise.resolve();
}).then(function () {
  expect(self.stub.calls.map(call => call.args[0])).to.deep.equal(expectations[dialect] || expectations.all});
});

sushantdhiman commented:
I have re-factored test to check for sequence of queries. Its ready for merge

janmeier commented:
LGTM, merge is awaiting the ES6 rewrite :)

janmeier commented on test/unit/transaction.test.js:
```diff
@@ -0,0 +1,53 @@
+      ],
+      mysql: [
+        'START TRANSACTION;',
+        'SET autocommit = 1;'
```
@sushantdhiman Actually, I'm not sure why this is needed? As I read the mysql docs 
> with autocommit mode enabled. This means that as soon as you execute a statement that updates (modifies) a table, MySQL stores the update on disk to make it permanent. The change cannot be rolled back.

This is the normal mode, outside of transactions.


>To disable autocommit mode implicitly for a single series of statements, use the START TRANSACTION statement:

http://dev.mysql.com/doc/refman/5.7/en/commit.html

So disabling autocommit with start transaction, then immediately enabling it again feels wrong.

sushantdhiman replied:
I'll check what happens when we disable auto commit in MySQL, may be related to #3568

sushantdhiman replied:
Hi @janmeier @mickhansen 

I have been investigating the reason why MySQL required `autocommit`. @janmeier is correct, as per `MySQL` docs we don't need to set it for `START TRANSACTION` which implies autocommit is set to false. 

But there is a bug in sequelize as noted in #3568, connection which use the `autocommit off` once will remain `off` until user manually doesn't switch it back `on`. This should happen inside `commit()` and `rollback()` automatically?

So can we merge this PR with MySQL `autocommit` set to 1 by default, as its supposed to? Then I will work on #3568 later on and will send a PR to fix it.

janmeier replied:
Agreed :)

sushantdhiman commented:
All green and ready for merge @janmeier
